contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryBase2_3.sol

```diff
@@ -675,66 +699,97 @@ abstract contract AutomationRegistryBase2_3 is ConfirmedOwner {
+    BillingConfig memory config = s_billingConfigs[billingToken];
+    params.flatFeeMicroLink = config.flatFeeMicroLink;
+    params.gasFeePPB = config.gasFeePPB;
+    (, int256 feedValue, , uint256 timestamp, ) = config.priceFeed.latestRoundData();
+    if (
+      feedValue <= 0 ||
+      block.timestamp < timestamp ||
+      (hotVars.stalenessSeconds > 0 && hotVars.stalenessSeconds < block.timestamp - timestamp)
+    ) {
+      params.priceUSD = config.fallbackPrice;
+    } else {
+      params.priceUSD = uint256(feedValue);
+    }
+    return params;
+  }
+
   /**
    * @dev calculates LINK paid for gas spent plus a configure premium percentage
    * @param hotVars the hot path variables
    * @param paymentParams the pricing data and gas usage data
-   * @dev use of PaymentParams is solely to avoid stack too deep errors
+   * @return receipt the receipt of payment with pricing breakdown
+   * @dev use of PaymentParams struct is necessary to avoid stack too deep errors
    */
   function _calculatePaymentAmount(
     HotVars memory hotVars,
     PaymentParams memory paymentParams
-  ) internal view returns (uint96, uint96) {
+  ) internal view returns (PaymentReceipt memory receipt) {
     uint256 gasWei = paymentParams.fastGasWei * hotVars.gasCeilingMultiplier;
     // in case it's actual execution use actual gas price, capped by fastGasWei * gasCeilingMultiplier
     if (paymentParams.isTransaction && tx.gasprice < gasWei) {
       gasWei = tx.gasprice;
     }
-    uint256 gasPayment = ((gasWei * (paymentParams.gasLimit + paymentParams.gasOverhead) + paymentParams.l1CostWei) *
-      paymentParams.nativeUSD) / paymentParams.linkUSD;
-    uint256 premium = (((gasWei * paymentParams.gasLimit) + paymentParams.l1CostWei) *
-      hotVars.paymentPremiumPPB *
-      paymentParams.nativeUSD) /
-      (paymentParams.linkUSD * 1e9) +
-      uint256(hotVars.flatFeeMicroLink) *
-      1e12;
-    // LINK_TOTAL_SUPPLY < UINT96_MAX
-    if (gasPayment + premium > LINK_TOTAL_SUPPLY) revert PaymentGreaterThanAllLINK();
-    return (uint96(gasPayment), uint96(premium));
+
+    uint256 gasPaymentUSD = (gasWei * (paymentParams.gasLimit + paymentParams.gasOverhead) + paymentParams.l1CostWei) *
+      paymentParams.nativeUSD; // this is USD * 1e36 ??? TODO
+    receipt.gasChargeWei = SafeCast.toUint96(gasPaymentUSD / paymentParams.billingToken.priceUSD);
+    receipt.gasReimbursementJuels = SafeCast.toUint96(gasPaymentUSD / paymentParams.linkUSD);
+
+    uint256 flatFeeUSD = uint256(paymentParams.billingToken.flatFeeMicroLink) * 1e12 * paymentParams.linkUSD; // TODO - this should get replaced by flatFeeCents later
+    uint256 premiumUSD = ((((gasWei * paymentParams.gasLimit) + paymentParams.l1CostWei) *
+      paymentParams.billingToken.gasFeePPB *
+      paymentParams.nativeUSD) / 1e9) + flatFeeUSD; // this is USD * 1e18
+    receipt.premiumWei = SafeCast.toUint96(premiumUSD / paymentParams.billingToken.priceUSD);
```

Comment:
Curious, this billingToken can be non-native, how we get Wei in the end?

Reply:
sorry, wei is probably a bad name. I will adjust it. I was using wei to mean "smallest denomination of billing token" but there isn't really a good word for that. I'm going to rename to just "premium".